File: 백준/Silver/2529. 부등호/부등호.py

global maxv 
global minv 
global smaxv
global sminv 
global N 

N = int(input())

operands = list(input().split()) 
maxv = -float('inf') 
minv = float('inf') 


def backtrack(ix, arr): 
    global maxv 
    global minv 
    global smaxv
    global sminv 


    if len(arr) == N+1: 
        sv = ''.join(arr)
        
        v = int(''.join(arr))
        if v > maxv:
            maxv = v 
            smaxv = sv 
        if v < minv: 
            minv  =v
            sminv = sv 
        return 


    for n in range(0,10): 
        
        if len(arr)==0: 
            backtrack(ix, arr+[str(n)] ) 
        else: 
            if operands[ix] =="<": 
                if int(arr[-1]) < n and str(n) not in arr: 
                    backtrack(ix+1, arr+[str(n)])
            else:  # ">"
                if int(arr[-1]) > n and str(n) not in arr: 
                    backtrack(ix+1 , arr+[str(n)])

backtrack(0,[])
print(smaxv)
print(sminv) 

File: 백준/Silver/2529. 부등호/부등호.cc
#include <iostream>
#include <string>

using namespace std; 

int N ; 
char operands[9] ; 
bool record[10] = {false ,} ; 
string maxv , minv ; 

bool check(int i , int j , char iq){
    if( iq == '<' ) return i < j ;
    else return i > j ; 
}

void backtrack(int ix , string s ){

    if( ix ==N+1){
        if(!minv.size()) minv = s; 
        else maxv = s; 
        return ; 
    }

    for(int n=0; n<10 ; n++){
        if( record[n]) continue ; 
        if( ix ==0 || check(s[ix-1] , n+'0', operands[ix-1])){
            record[n] = true ; 
            backtrack(ix+1 , s+to_string(n));
            record[n] = false ;
        }
    }
}

int main() { 

    cin >> N ; 
    int t = N ; 
    for(int i=0; i < N ; i++) cin >> operands[i] ; 
    
    backtrack(0, "") ; 
    cout << maxv << '\n' ; 
    cout << minv << '\n' ; 
    return 0 ; 

}
